### PR TITLE
Improve memory read/write

### PR DESCRIPTION
**Pull request by qwe661234**
After analyzing the data collected during the execution of the Dhrystone benchmark, we found that the primary performance bottleneck lies in memory read/write operations. To address this issue, we modified the implementation of memory read/write by eliminating unnecessary checks and replacing slow operations with more efficient ones. Additionally, we integrated a simple memory pool and resolved the issue of disabled unaligned memory access, the user can add argument `--misalign` to enable misaligned memory access.

After analyzing the performance results from running Dhrystone, we observed a significant improvement following our enhancements to the memory I/O. 

| Test | 66200d0 | improvement | Speedup |
|--------| -------- | -------- |-------- |
| Dhrystone  | 815 DMIPS    |  1245 DMIPS | +52.7% |

See: #122

### CODE REVIEW

File: src/riscv_private.h [hunk 1]
```diff
@@ -119,6 +119,8 @@ struct riscv_internal {
 
     bool compressed;       /**< current instruction is compressed or not */
     block_map_t block_map; /**< basic block map */
+
+    bool is_misaliged_enabled; /**< enable memory misaligned access */
```
Comment (jserv) [inline src/riscv_private.h]: Rename to `allow_ misalign`.

**Review by jserv — CHANGES_REQUESTED** [review]
Since the misaligned memory access is optional, we don't have to check it for the internals of emulator. Instead, we can register the callback functions while initializing the emulator.

See [async.h](https://github.com/embedded2016/server-framework/blob/master/async.h) and [async.c](https://github.com/embedded2016/server-framework/blob/master/async.c) for details.
```c
/* API gateway */
struct __ASYNC_API__ Async = {
    .create = async_create,
    .signal = async_signal,
    .wait = async_wait,
    .finish = async_finish,
    .run = async_run,
};
```

Once you prepare the functions specialized for both misaligned memory and strictly-aligned memory access, the proper I/O implementations can be hooked up. Consequently, you don't have to check `allow_ misalign` all the time.

File: src/io.c [hunk 2]
```diff
@@ -125,14 +109,46 @@ void memory_write(memory_t *m, uint32_t addr, const uint8_t *src, uint32_t size)
...
         c->data[p & 0xffff] = src[i];
     }
 }
 
+void memory_write_w(memory_t *m, uint32_t addr, const uint8_t *src)
+{
+    const uint32_t addr_lo = addr & mask_lo, addr_hi = addr >> 16;
+    chunk_t *c = m->chunks[addr_hi];
+    if (unlikely(!c)) {
+        c = calloc(1, sizeof(chunk_t));
```
Comment (jserv) [inline src/io.c]: Can you eliminate the use of `calloc` here?

File: src/main.c [hunk 3]
```diff
@@ -202,7 +211,7 @@ int main(int argc, char **args)
         state->break_addr = end->st_value;
 
     /* create the RISC-V runtime */
-    riscv_t *rv = rv_create(&io, state);
+    riscv_t *rv = rv_create(&io, state, opt_misaligned);
```
Comment (jserv) [inline src/main.c]: Don't change the prototype of `rv_create()`. Instead, consolidate `state` or `io`.

**Comment by jserv** [comment]
> | Test	| 66200d0	| Proposed |	Speedup |
> |-------|--------------|-----------|-------------|
> | Dhrystone	| 815 DMIPS	| 1200 DMIPS	| +47.2% |

What else do you consider it would be beneficial to employ in place of Spike for faster interpreting?

File: src/io.c (hunk 2 above)
Comment (qwe661234) [reply src/io.c]: If we want to eliminate it, we should know the total memory size and allocate it first.

**Comment by qwe661234** [comment]
> Since the misaligned memory access is optional, we don't have to check it for the internals of emulator. Instead, we can register the callback functions while initializing the emulator.
> 
> See [async.h](https://github.com/embedded2016/server-framework/blob/master/async.h) and [async.c](https://github.com/embedded2016/server-framework/blob/master/async.c) for details.
> 
> ```c
> /* API gateway */
> struct __ASYNC_API__ Async = {
>     .create = async_create,
>     .signal = async_signal,
>     .wait = async_wait,
>     .finish = async_finish,
>     .run = async_run,
> };
> ```
> 
> Once you prepare the functions specialized for both misaligned memory and strictly-aligned memory access, the proper I/O implementations can be hooked up. Consequently, you don't have to check `allow_ misalign` all the time.

File: src/io.c (hunk 2 above)
Comment (jserv) [reply src/io.c]: Let's introduce memory pool within the pull request.

**Comment by qwe661234** [comment]
> Since the misaligned memory access is optional, we don't have to check it for the internals of emulator. Instead, we can register the callback functions while initializing the emulator.
> 
> See [async.h](https://github.com/embedded2016/server-framework/blob/master/async.h) and [async.c](https://github.com/embedded2016/server-framework/blob/master/async.c) for details.
> 
> ```c
> /* API gateway */
> struct __ASYNC_API__ Async = {
>     .create = async_create,
>     .signal = async_signal,
>     .wait = async_wait,
>     .finish = async_finish,
>     .run = async_run,
> };
> ```
> 
> Once you prepare the functions specialized for both misaligned memory and strictly-aligned memory access, the proper I/O implementations can be hooked up. Consequently, you don't have to check `allow_ misalign` all the time.

But the memory I/O implemention for for both misaligned memory and strictly-aligned memory access is the same, the only difference is that we need to handle exception if misaligned memory access is disabled.

**Comment by jserv** [comment]
> the memory I/O implemention for for both misaligned memory and strictly-aligned memory access is the same, the only difference is that we need to handle exception if misaligned memory access is disabled.

Then, you can hook exception handlers which are aware of misaligned memory access by means of runtime dispatch.

**Comment by qwe661234** [comment]
> > the memory I/O implemention for for both misaligned memory and strictly-aligned memory access is the same, the only difference is that we need to handle exception if misaligned memory access is disabled.
> 
> Then, you can hook exception handlers which are aware of misaligned memory access by means of runtime dispatch.

Does this means send a callback function to memory read/ write? Because we don't need load/store handler if the memory misaligned access is enabled.

**Comment by jserv** [comment]
> > Then, you can hook exception handlers which are aware of misaligned memory access by means of runtime dispatch.
> Does this means send a callback function to memory read/ write?

Sort of. In particular, we can shrink the code below:
```c
/* SW: Store Word */
RVOP(sw, {
    const uint32_t addr = rv->X[ir->rs1] + ir->imm;
    if (!rv->is_misaliged_enabled && unlikely(addr & 3)) {
        rv->compressed = false;
        rv_except_store_misaligned(rv, addr);
        return false;
    ...
```
by replacing with something like the following:
```c
/* SW: Store Word */
RVOP(sw, {
    const uint32_t addr = rv->X[ir->rs1] + ir->imm;
    RV_EXC_MISAGLIN_HANDLER(); /* some macro */
```

Meanwhile, we can hide the implementation details about `allow_ misalign` checks.

File: src/riscv.h [hunk 4]
```diff
@@ -107,6 +107,9 @@ typedef struct {
     /* system */
     riscv_on_ecall on_ecall;
     riscv_on_ebreak on_ebreak;
+
+    /* enable memory misaligned access */
```
Comment (jserv) [inline src/riscv.h]: Should be "enable misaligned memory access"

File: src/emulate.c [hunk 5]
```diff
@@ -90,6 +90,18 @@ static void rv_exception_default_handler(riscv_t *rv)
 RV_EXCEPTION_LIST
 #undef _
 
+#define RV_EXC_MISAGLIN_HANDLER(mask_or_pc, type, compress, IO)    \
```
Comment (jserv) [inline src/emulate.c]: Typo. Should be `RV_EXC_MISALIGN_HANDLER`.

Comment (jserv) [inline src/emulate.c]: Add some comments for this macro.

File: src/emulate.c [hunk 6]
```diff
@@ -90,6 +90,19 @@ static void rv_exception_default_handler(riscv_t *rv)
 RV_EXCEPTION_LIST
 #undef _
 
+/* wrap load/store and insn misaligned handler */
```
Comment (jserv) [inline src/emulate.c]: You shall explain the parameters in details.

**Review by jserv — CHANGES_REQUESTED** [review]
Integrate memory pool in conjunction of proposed memory operation changes.

File: src/emulate.c (hunk 6 above)
Comment (qwe661234) [reply src/emulate.c]: Updated

File: src/emulate.c [hunk 7]
```diff
@@ -90,6 +90,25 @@ static void rv_exception_default_handler(riscv_t *rv)
 RV_EXCEPTION_LIST
 #undef _
 
+/* wrap load/store and insn misaligned handler
+ * @mask_or_pc: misaligned mask for load/store misaligned handler and pc for
```
Comment (jserv) [inline src/emulate.c]: Shorten the description for `mask_or_pc`.

File: src/io.c (hunk 2 above)
Comment (jserv) [reply src/io.c]: You don't have to integrate TLSF (or its variants such as xvmalloc) within this commit. Instead, an arbitrary memory pool works for this case.

Comment (qwe661234) [reply src/io.c]: I roughly integrate the [TLSF 2.4.6](http://www.gii.upv.es/tlsf/main/repo) to file `io.c`.

**Review by jserv — CHANGES_REQUESTED** [review]
Get rid of GPL'd TLSF. You MUST use the BSD license re-implementation instead. See #94 .
Original TLSF is licensed under GNU GPL, which raises more restrictions over MIT License. We can integrate BSD License components in rv32emu, but it is not allowed to relicense GPL'd ones into rv32emu. That is exactly what BSD re-implementation appears.

**Comment by jserv** [comment]
Also, the original TLSF suffers from several security concerns. For this issue, we only need a simplified memory pool without integrating TLSF, where #94 emphasizes.

**Comment by qwe661234** [comment]
> Also, the original TLSF suffers from several security concerns. For this issue, we only need a simplified memory pool without integrating TLSF, where #94 emphasizes.

I believe that the memory pool with a fixed chunk size can fulfill our requirements, so I decided to integrate this type of memory pool into the file `io.c`.

**Review by jserv — CHANGES_REQUESTED** [review]
Squash the commits along with comprehensive git commit messages. In addition, show the latest benchmark results.

File: src/emulate.c [hunk 8]
```diff
@@ -90,6 +90,24 @@ static void rv_exception_default_handler(riscv_t *rv)
 RV_EXCEPTION_LIST
 #undef _
 
+/* wrap load/store and insn misaligned handler
+ * @mask_or_pc: mask for load/store and pc for insn misaligned handler.
+ * @type: type of misaligned handler
+ * @compress: compressed instruction or not
+ * @IO: misaligned handler is load/store or insn
```
Comment (jserv) [inline src/emulate.c]: Improve the writing for `@IO` entry.

Comment (jserv) [reply src/emulate.c]: Ask ChatGPT to proofread.

File: src/io.c [hunk 9]
```diff
@@ -9,27 +9,25 @@
 #include <string.h>
 
 #include "io.h"
+#include "mpool.h"
 
 static const uint32_t mask_lo = 0xffff;
 static const uint32_t mask_hi = ~(0xffff);
 
+struct mpool *mp;
```
[the remark this answers is not in the document]
Comment (jserv) [reply src/io.c]: Can you make it local? That is, declare this variable as `static`.

File: src/io.c [hunk 10]
```diff
@@ -125,23 +107,54 @@ void memory_write(memory_t *m, uint32_t addr, const uint8_t *src, uint32_t size)
         uint32_t x = p >> 16;
         chunk_t *c = m->chunks[x];
         if (!c) {
-            c = malloc(sizeof(chunk_t));
-            memset(c->data, 0, sizeof(c->data));
+            c = calloc(1, sizeof(chunk_t));
```
Comment (jserv) [inline src/io.c]: The region should be requested from memory pool.

File: src/mpool.c [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+typedef struct mpool {
+    long chunk_ct;
+    long page_count;
+    size_t chunk_size;
+    char *mempage;
+    struct memchunk *free_chunk_head;
+    struct mpool *next;
+} mpool_t;
+
+static void *get_mmap(long sz)
```
Comment (jserv) [inline src/mpool.c]: Check the file `src/elf.c` for `USE_MMAP`. At present, you can figure out Microsoft Windows counterpart or stop compilation (by placing `#error`).

Portability is an important consideration.

File: src/mpool.h [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+ */
+
+#include <assert.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <unistd.h>
+
+struct memchunk;
```
Comment (jserv) [inline src/mpool.h]: `struct memchunk` is implementation specific. Don't appear in this header file.

File: src/mpool.h [hunk 13]
```diff
@@ -0,0 +1,24 @@
+/*
+ * rv32emu is freely redistributable under the MIT License. See the file
+ * "LICENSE" for information on usage and redistribution of this file.
+ */
+
```
Comment (jserv) [inline src/mpool.h:6]: Add proper `#pragma once`.

File: src/mpool.h [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+#include <assert.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/mman.h>
```
Comment (jserv) [inline src/mpool.h]: Minimize the header inclusion! Only keep the necessary ones.

File: src/mpool.c [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+ */
+
+#include "mpool.h"
+
+typedef struct memchunk {
+    struct memchunk *next;
+} memchunk_t;
+
+typedef struct mpool {
+    long chunk_ct;
```
Comment (jserv) [inline src/mpool.c]: `chunk_ct` is not a appropriate mnemonic.

File: src/mpool.c [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+    long page_count;
+    size_t chunk_size;
+    char *mempage;
```
Comment (jserv) [inline src/mpool.c]: Rename it to `mapped`, which stands for mapped memory region.

File: src/mpool.c [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+#include "mpool.h"
+
+typedef struct memchunk {
+    struct memchunk *next;
+} memchunk_t;
+
+typedef struct mpool {
+    long chunk_ct;
+    long page_count;
+    size_t chunk_size;
```
Comment (jserv) [inline src/mpool.c:27]: Check if `size_t` fits the range requirements for `chunk_ct` and `page_count`.

File: src/mpool.c [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+    void *p =
+        mmap(0, sz, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANON, -1, 0);
+    if (p == MAP_FAILED)
+        return NULL;
+    return p;
+}
+
+mpool_t *mpool_create(size_t pool_size, size_t chunk_size)
+{
+    mpool_t *new_mp = malloc(sizeof(mpool_t));
```
Comment (jserv) [inline src/mpool.c:49]: Return `NULL` when `malloc` or `mmap` fails.

File: src/io.c [hunk 19]
```diff
@@ -125,23 +107,54 @@ void memory_write(memory_t *m, uint32_t addr, const uint8_t *src, uint32_t size)
...
             m->chunks[x] = c;
         }
         c->data[p & 0xffff] = src[i];
     }
 }
 
+void memory_write_w(memory_t *m, uint32_t addr, const uint8_t *src)
```
Comment (jserv) [inline src/io.c]: Use macro to generate the function bodies of `memory_write_{w,s,b}`.

**Review by jserv — CHANGES_REQUESTED** [review]
Append `Close #122` at the end of git commit messages.

File: src/mpool.h [hunk 20]
```diff
@@ -0,0 +1,25 @@
+/*
+ * rv32emu is freely redistributable under the MIT License. See the file
+ * "LICENSE" for information on usage and redistribution of this file.
+ */
+#pragma once
+
+#include <stdlib.h>
+#include <string.h>
+#if !defined(_WIN32)
```
Comment (jserv) [inline src/mpool.h]: No, implementation specific statements should no appear in this header! Keep it simple and stupid.

File: src/mpool.h [hunk 21]
```diff
@@ -0,0 +1,20 @@
+/*
+ * rv32emu is freely redistributable under the MIT License. See the file
+ * "LICENSE" for information on usage and redistribution of this file.
+ */
+#pragma once
+
+#include <stdlib.h>
```
Comment (jserv) [inline src/mpool.h]: Only `<stddef.h>` should be included here.

File: src/mpool.c [hunk 22]
```diff
@@ -0,0 +1,109 @@
...
+typedef struct mpool {
+    size_t chunk_count;
+    size_t page_count;
+    size_t chunk_size;
+    char *mapped;
+    struct memchunk *free_chunk_head;
+    struct mpool *next;
+} mpool_t;
+
+static void *get_mmap(size_t sz)
```
Comment (jserv) [inline src/mpool.c]: Rename to `mem_arena` and add comments.

File: src/io.c [hunk 23]
```diff
@@ -9,27 +9,25 @@
...
 static const uint32_t mask_hi = ~(0xffff);
 
+static struct mpool *mp;
+
 memory_t *memory_new()
 {
-    memory_t *m = malloc(sizeof(memory_t));
-    memset(m->chunks, 0, sizeof(m->chunks));
+    memory_t *m = calloc(1, sizeof(memory_t));
+    mp = mpool_create(sizeof(chunk_t) << 5, sizeof(chunk_t));
```
Comment (jserv) [inline src/io.c]: You shall explain why shift 5 is here.

File: src/io.c [hunk 24]
```diff
@@ -85,63 +83,61 @@ uint32_t memory_ifetch(memory_t *m, uint32_t addr)
...
-            m->chunks[x] = c;
+            c = mpool_calloc(mp);
+            m->chunks[addr_hi] = c;
         }
-        c->data[p & 0xffff] = src[i];
+        c->data[addr_lo] = src[i];
     }
 }
 
+#define memory_write(size, type)                                             \
```
Comment (jserv) [inline src/io.c]: Rename to `MEM_WRITE_IMPL`.

**Review by jserv — CHANGES_REQUESTED** [review]
Rephrase the git commit message as following:
> After analyzing the data collected during the execution of the Dhrystone benchmark, it was discovered that the primary performance bottleneck lies in memory read/write operations. To address this issue, modifications were made to the implementation of memory read/write operations, eliminating unnecessary checks and replacing slow operations with more efficient ones. Additionally, a simple memory pool was integrated. It is now possible for the user to enable misaligned memory access by launching with option "--misalign".
> 
> The enhancements made to the memory I/O resulted in a significant improvement, as observed through performance results obtained by running Dhrystone.

File: src/mpool.h (hunk 21 above)
Comment (qwe661234) [reply src/mpool.h]: The gcc-12 compiler shows the warning when I only include `<stddef.h>`.

Comment (jserv) [reply src/mpool.h]: I don't think so.
```
gcc-12 -c mpool.h -Wall
```
It compiles without warnings. The actual header inclusion should appear in C source files.

File: src/mpool.h [hunk 25]
```diff
@@ -0,0 +1,20 @@
...
+#include <string.h>
+
+struct mpool;
+
+struct mpool *mpool_create(size_t pool_size, size_t chunk_size);
```
Comment (jserv) [inline src/mpool.h:17]: Add informative comments for each function.

File: src/io.c [hunk 26]
```diff
@@ -9,27 +9,27 @@
...
 static const uint32_t mask_hi = ~(0xffff);
 
+static struct mpool *mp;
+
 memory_t *memory_new()
 {
-    memory_t *m = malloc(sizeof(memory_t));
-    memset(m->chunks, 0, sizeof(m->chunks));
+    memory_t *m = calloc(1, sizeof(memory_t));
+    /* sizeof(chunk_t) << 5 satisfies the memory requirement for all test cases
```
Comment (jserv) [inline src/io.c]: If `sizeof(chunk_t) << 5` does not fit, there should be some mechanism to extend.

File: src/mpool.c [hunk 27]
```diff
@@ -0,0 +1,109 @@
+/*
+ * rv32emu is freely redistributable under the MIT License. See the file
+ * "LICENSE" for information on usage and redistribution of this file.
+ */
+#if !defined(_WIN32)
+#define USE_MMAP 1
+#include <sys/mman.h>
+#include <unistd.h>
+#endif
+
```
Comment (jserv) [inline src/mpool.c:12]: Add proper header inclusion here.

File: src/emulate.c [hunk 28]
```diff
@@ -90,6 +90,24 @@ static void rv_exception_default_handler(riscv_t *rv)
...
+ * @compress: compressed instruction or not
+ * @IO: whether the misaligned handler is for load/store or insn.
+ */
+#define RV_EXC_MISALIGN_HANDLER(mask_or_pc, type, compress, IO)    \
+    IIF(IO)                                                        \
+    (if (!rv->io.allow_misalign && unlikely(addr & (mask_or_pc))), \
+     if (unlikely(insn_is_misaligned(rv->PC))))                    \
+    {                                                              \
+        rv->compressed = compress;                                 \
+        IIF(IO)                                                    \
```
Comment (jserv) [inline src/emulate.c]: Can you simplify as following?
```c
rv_except_##type##_misaligned(rv, IIF(IO)(addr, mask_or_pc));
```

**Review by jserv — CHANGES_REQUESTED** [review]
Append a test item which triggers option `--misalign` in GitHub Actions. That is, we shall regularly validate the case when misaligned memory access is allowed for CI pipeline.

**Pull request merged**